Lambdify with cse wrong result

Using `lambdify` with `cse=True` sometimes returns the wrong result.
Minimal example:
```python
import sympy as sp
c = sp.symbols("c")
x = sp.symbols("x0", real=True)
poly = c*x**2 + c*(x**4 - x**2)

f = sp.lambdify((c, x), poly, cse=True)  #  <- `cse=True` makes the difference
g = sp.lambdify((c, x), poly, cse=False)

#  wrong result, AssertionError: 256 /= 16
assert f(1, 2) == g(1, 2), f"{f(1, 2)} /= {g(1, 2)}"
```

**Cause**
The issue seems to be related to the filtering of symbols to be excluded in the `tree_cse` function around here [https://github.com/sympy/sympy/blob/master/sympy/simplify/cse_main.py#L655](https://github.com/sympy/sympy/blob/53a36aa71956c988e2b6ce33db0d8cbcbdfacfbb/sympy/simplify/cse_main.py#L655).

From what I can tell, the cse method aims to avoid name collisions between symbols already in the expression and the symbols representing common subexpressions.
This is done by adding all symbols found in the expression to an `excluded_symbols` set and filtering the `symbols` iterator to exclude them.
However, symbols are equal iff their names and assumptions match, so this approach does not avoid symbol *name collisions*:

```python
from sympy import Symbol
Symbol("x") in {Symbol("x", real=True)}  # returns False
```

**Suggested solution**
In the `tree_cse` function, instead of filtering `symbols` from all symbols occuring in the expression, 
filter from all symbols sharing `name` with some symbol in the expression.

If this can be considered a sufficient fix for the issue I'd be happy to provide a PR.

--- a/.mailmap
+++ b/.mailmap
@@ -747,6 +747,7 @@ Johan Blåbäck <johan_bluecreek@riseup.net> <johan.blaback@cea.fr>
 Johan Guzman <jguzm022@ucr.edu>
 Johann Cohen-Tanugi <johann.cohentanugi@gmail.com>
 Johannes Hartung <joha2@gmx.net>
+Johannes Kasimir <johannes.kasimir@math.lu.se>
 John Connor <john.theman.connor@gmail.com>
 John Möller <john.moller@outlook.com>
 John V. Siratt <jvsiratt@gmail.com>

--- a/sympy/simplify/cse_main.py
+++ b/sympy/simplify/cse_main.py
@@ -621,7 +621,7 @@ def tree_cse(exprs, symbols, opt_subs=None, order='canonical', ignore=()):
                 expr.is_Order or
                 isinstance(expr, (MatrixSymbol, MatrixElement))):
             if expr.is_Symbol:
-                excluded_symbols.add(expr)
+                excluded_symbols.add(expr.name)
             return
 
         if iterable(expr):
@@ -652,7 +652,7 @@ def tree_cse(exprs, symbols, opt_subs=None, order='canonical', ignore=()):
     ## Rebuild tree
 
     # Remove symbols from the generator that conflict with names in the expressions.
-    symbols = (symbol for symbol in symbols if symbol not in excluded_symbols)
+    symbols = (_ for _ in symbols if _.name not in excluded_symbols)
 
     replacements = []
 

--- a/sympy/simplify/tests/test_cse.py
+++ b/sympy/simplify/tests/test_cse.py
@@ -246,6 +246,14 @@ def test_issue_6263():
     assert cse(e, optimizations='basic') == ([], [True])
 
 
+def test_issue_25043():
+    c = symbols("c")
+    x = symbols("x0", real=True)
+    cse_expr = cse(c*x**2 + c*(x**4 - x**2))[-1][-1]
+    free = cse_expr.free_symbols
+    assert len(free) == len({i.name for i in free})
+
+
 def test_dont_cse_tuples():
     from sympy.core.function import Subs
     f = Function("f")

